README.md
# Circular Text Widget

[![pub package](https://img.shields.io/pub/v/flutter_circular_text.svg)](https://pub.dartlang.org/packages/flutter_circular_text)

## Installation

Add dependency in `pubspec.yaml`:
```yaml
dependencies:
  flutter_circular_text: "^0.3.1"
```

Import in your project:
```dart
import 'package:flutter_circular_text/circular_text.dart';
```

## Basic usage

```dart
CircularText(
  children: [
    TextItem(
      text: Text(
        "Chuck Norris".toUpperCase(),
        style: TextStyle(
          fontSize: 28,
          color: Colors.blue,
          fontWeight: FontWeight.bold,
        ),
      ),
      space: 12,
      startAngle: -90,
      startAngleAlignment: StartAngleAlignment.center,
      direction: CircularTextDirection.clockwise,
    ),
    TextItem(
      text: Text(
        "top 100 Facts".toUpperCase(),
        style: TextStyle(
          fontSize: 20,
          color: Colors.amberAccent,
          fontWeight: FontWeight.bold,
        ),
      ),
      space: 10,
      startAngle: 90,
      startAngleAlignment: StartAngleAlignment.center,
      direction: CircularTextDirection.anticlockwise,
    ),
  ],
  radius: 125,
  position: CircularTextPosition.inside,
  backgroundPaint: Paint()..color = Colors.grey.shade200,
)
```

## Examples

[example](https://github.com/faob-dev/flutter_circular_text/tree/master/example) project contains single and multi text demos

### Demos

##### Single Text Demo
![alt tag](https://raw.githubusercontent.com/faob-dev/flutter_circular_text/master/screenshots/single_circular_text.gif)

##### Multi Text Demo
<img src="https://raw.githubusercontent.com/faob-dev/flutter_circular_text/master/screenshots/multi_circular_text.jpeg" width=270>

## Changelog
Check [Changelog](https://github.com/faob-dev/flutter_circular_text/blob/master/CHANGELOG.md) for updates

## Bugs/Requests
Reporting issues and requests for new features are always welcome.


CHANGELOG.md
## 0.3.3
* bugfix: correct printing of unicode/emoji chars

## 0.3.2
* bugfix: avoid an exception when adding a new `TextItem`

## 0.3.1
* Added null safety

## 0.3.0
* Added multi `Text` support
* Added `StartAngleAlignment` property to align text around `startAngle`
* Added single and multi text examples

## 0.2.1
* Added `Text`, changed angle position and added some minor changes

## 0.2.0
* Added support to handle `PaintingStyle.stroke`

## 0.1.0
* Initial version

example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter_test/flutter_test.dart';


void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
//    await tester.pumpWidget(MyApp());
//
    // Verify that our counter starts at 0.
//    expect(find.text('0'), findsOneWidget);
//    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
//    await tester.tap(find.byIcon(Icons.add));
//    await tester.pump();

    // Verify that our counter has incremented.
//    expect(find.text('0'), findsNothing);
//    expect(find.text('1'), findsOneWidget);
  });
}


lib/circular_text/model.dart
import 'package:flutter/material.dart';

enum CircularTextDirection { clockwise, anticlockwise }

enum CircularTextPosition { outside, inside }

enum StartAngleAlignment { start, center, end }

class TextItem {
  /// Text
  final Text text;

  /// Space between characters
  final double space;

  /// Text starting position
  final double startAngle;

  /// Text alignment around [startAngle]
  /// [StartAngleAlignment.start] text will starts from [startAngle]
  /// [StartAngleAlignment.center] text will be centered on [startAngle]
  /// [StartAngleAlignment.end] text will ends on [startAngle]
  final StartAngleAlignment startAngleAlignment;

  /// Text direction either clockwise or anticlockwise
  final CircularTextDirection direction;

  TextItem({
    required this.text,
    this.space = 10,
    this.startAngle = 0,
    this.startAngleAlignment = StartAngleAlignment.start,
    this.direction = CircularTextDirection.clockwise,
  }) : assert(space >= 0);

  bool isChanged(TextItem oldTextItem) {
    bool isTextChanged() {
      return oldTextItem.text.data != text.data ||
          oldTextItem.text.style != text.style;
    }

    return isTextChanged() ||
        oldTextItem.space != space ||
        oldTextItem.startAngle != startAngle ||
        oldTextItem.startAngleAlignment != startAngleAlignment ||
        oldTextItem.direction != direction;
  }
}


lib/circular_text/widget.dart
import 'dart:math';

import 'package:flutter/material.dart';

import 'model.dart';

class CircularText extends StatelessWidget {
  ///List of text
  final List<TextItem> children;

  /// Circle radius
  final double radius;

  /// Text position either outside or inside circle
  final CircularTextPosition position;

  /// Background paint
  final Paint? backgroundPaint;

  CircularText({
    Key? key,
    required this.children,
    this.radius = 125,
    this.position = CircularTextPosition.inside,
    this.backgroundPaint,
  })  : assert(radius >= 0),
        super(key: key);

  @override
  Widget build(BuildContext context) {
    return FittedBox(
      child: SizedBox.fromSize(
        size: Size(2 * radius, 2 * radius),
        child: CustomPaint(
          painter: _CircularTextPainter(
            children: children,
            position: position,
            backgroundPaint: backgroundPaint,
            textDirection: Directionality.of(context),
          ),
        ),
      ),
    );
  }
}

class _CircularTextPainter extends CustomPainter {
  final List<TextItem> children;
  final CircularTextPosition position;
  final Paint backgroundPaint;
  final TextDirection textDirection;

  double _radius = 0.0;

  _CircularTextPainter({
    required this.children,
    this.position = CircularTextPosition.inside,
    Paint? backgroundPaint,
    required this.textDirection,
  }) : this.backgroundPaint =
            backgroundPaint ?? (Paint()..color = Colors.transparent);

  @override
  void paint(Canvas canvas, Size size) {
    _radius = min(size.width / 2, size.height / 2);
    canvas.translate(size.width / 2, size.height / 2);
    canvas.drawCircle(Offset.zero, _radius, backgroundPaint);

    for (final textItem in children) {
      canvas.save();
      List<TextPainter> _charPainters = [];
      Text text = textItem.text;
      for (final int rune in text.data!.runes) {
        _charPainters.add(TextPainter(
            text: TextSpan(text: String.fromCharCode(rune), style: text.style),
            textDirection: textDirection)
          ..layout());
      }
      if (textItem.direction == CircularTextDirection.clockwise) {
        _paintTextClockwise(canvas, size, textItem, _charPainters);
      } else {
        _paintTextAntiClockwise(canvas, size, textItem, _charPainters);
      }
      canvas.restore();
    }
  }

  void _paintTextClockwise(Canvas canvas, Size size, TextItem textItem,
      List<TextPainter> _charPainters) {
    bool hasStrokeStyle = backgroundPaint.style == PaintingStyle.stroke &&
        backgroundPaint.strokeWidth > 0.0;

    double angleShift = _calculateAngleShift(textItem, _charPainters.length);
    canvas.rotate((textItem.startAngle + 90 - angleShift) * pi / 180);
    for (int i = 0; i < _charPainters.length; i++) {
      final tp = _charPainters[i];
      final x = -tp.width / 2;
      final y = position == CircularTextPosition.outside
          ? (-_radius - tp.height) -
              (hasStrokeStyle ? backgroundPaint.strokeWidth / 2 : 0.0)
          : -_radius - (hasStrokeStyle ? tp.height / 2 : 0.0);

      tp.paint(canvas, Offset(x, y));
      canvas.rotate(textItem.space * pi / 180);
    }
  }

  void _paintTextAntiClockwise(Canvas canvas, Size size, TextItem textItem,
      List<TextPainter> _charPainters) {
    bool hasStrokeStyle = backgroundPaint.style == PaintingStyle.stroke &&
        backgroundPaint.strokeWidth > 0.0;

    double angleShift = _calculateAngleShift(textItem, _charPainters.length);
    canvas.rotate((textItem.startAngle - 90 + angleShift) * pi / 180);
    for (int i = 0; i < _charPainters.length; i++) {
      final tp = _charPainters[i];
      final x = -tp.width / 2;
      final y = position == CircularTextPosition.outside
          ? _radius + (hasStrokeStyle ? backgroundPaint.strokeWidth / 2 : 0.0)
          : (_radius - tp.height) + (hasStrokeStyle ? tp.height / 2 : 0.0);

      tp.paint(canvas, Offset(x, y));
      canvas.rotate(-textItem.space * pi / 180);
    }
  }

  double _calculateAngleShift(TextItem textItem, int textLength) {
    double angleShift = -1;
    switch (textItem.startAngleAlignment) {
      case StartAngleAlignment.start:
        angleShift = 0;
        break;
      case StartAngleAlignment.center:
        int halfItemsLength = textLength ~/ 2;
        if (textLength % 2 == 0) {
          angleShift =
              ((halfItemsLength - 1) * textItem.space) + (textItem.space / 2);
        } else {
          angleShift = (halfItemsLength * textItem.space);
        }
        break;
      case StartAngleAlignment.end:
        angleShift = (textLength - 1) * textItem.space;
        break;
    }
    return angleShift;
  }

  @override
  bool shouldRepaint(_CircularTextPainter oldDelegate) {
    bool isTextItemsChanged() {
      bool isChanged = false;
      for (int i = 0; i < children.length; i++) {
        if (i >= oldDelegate.children.length || children[i].isChanged(oldDelegate.children[i])) {
          isChanged = true;
          break;
        }
      }
      return isChanged;
    }

    bool isBackgroundPaintChanged() {
      return oldDelegate.backgroundPaint.color != backgroundPaint.color ||
          oldDelegate.backgroundPaint.style != backgroundPaint.style ||
          oldDelegate.backgroundPaint.strokeWidth !=
              backgroundPaint.strokeWidth;
    }

    return isTextItemsChanged() ||
        oldDelegate.position != position ||
        oldDelegate.textDirection != textDirection ||
        isBackgroundPaintChanged();
  }
}


example/lib/single_text_example.dart
import 'package:flutter/material.dart';
import 'package:flutter_circular_text/circular_text.dart';

class SingleTextDemo extends StatefulWidget {
  @override
  _SingleTextDemoState createState() => _SingleTextDemoState();
}

class _SingleTextDemoState extends State<SingleTextDemo> {
  double _space = 10;
  double _startAngle = 0;
  double _strokeWidth = 0.0;
  bool _showStroke = false;
  bool _showBackground = true;
  StartAngleAlignment _startAngleAlignment = StartAngleAlignment.start;
  CircularTextPosition _position = CircularTextPosition.inside;
  CircularTextDirection _direction = CircularTextDirection.clockwise;

  @override
  Widget build(BuildContext context) {
    return Material(
      child: Container(
        color: Colors.white,
        alignment: Alignment.center,
        padding: EdgeInsets.only(left: 10, right: 10, top: 50, bottom: 10),
        child: SizedBox.fromSize(
          size: Size(360, double.infinity),
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            crossAxisAlignment: CrossAxisAlignment.center,
            children: <Widget>[
              buildCircularTextWidget(),
              SizedBox.fromSize(size: Size(40, 40)),
              Expanded(
                flex: 1,
                child: CustomScrollView(
                  slivers: <Widget>[
                    buildSpacePanel(),
                    buildStartAnglePanel(),
                    buildCircularTextAnglePositionPanel(),
                    buildStrokeWidthPanel(),
                    buildShowBackgroundShapePanel(),
                    buildCircularTextPositionPanel(),
                    buildCircularTextDirectionPanel(),
                  ],
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }

  Widget buildCircularTextWidget() {
    final backgroundPaint = Paint();
    if (_showBackground) {
      backgroundPaint..color = Colors.grey.shade200;
      if (_showStroke) {
        backgroundPaint
          ..color = Colors.grey.shade200
          ..style = PaintingStyle.stroke
          ..strokeWidth = _strokeWidth;
      }
    } else {
      backgroundPaint.color = Colors.transparent;
    }

    return CircularText(
      children: [
        TextItem(
          text: Text(
            "circular text".toUpperCase(),
            style: TextStyle(
              fontSize: 25,
              color: Colors.blue,
              fontWeight: FontWeight.bold,
            ),
          ),
          space: _space,
          startAngle: _startAngle,
          startAngleAlignment: _startAngleAlignment,
          direction: _direction,
        ),
      ],
      radius: 125,
      position: _position,
      backgroundPaint: backgroundPaint,
    );
  }

  Widget buildSpacePanel() {
    return SliverToBoxAdapter(
      child: Row(
        mainAxisAlignment: MainAxisAlignment.spaceBetween,
        children: <Widget>[
          Text("SPACE", style: TextStyle(fontWeight: FontWeight.bold)),
          Slider(
            value: _space,
            min: 0,
            max: 30,
            onChanged: (value) {
              setState(() => _space = value);
            },
          )
        ],
      ),
    );
  }

  Widget buildStartAnglePanel() {
    return SliverToBoxAdapter(
      child: Row(
        mainAxisAlignment: MainAxisAlignment.spaceBetween,
        children: <Widget>[
          Text("START ANGLE", style: TextStyle(fontWeight: FontWeight.bold)),
          Slider(
            value: _startAngle,
            min: 0,
            max: 360,
            divisions: 4,
            label: "${_startAngle.toInt()}",
            onChanged: (value) {
              setState(() => _startAngle = value);
            },
          )
        ],
      ),
    );
  }

  Widget buildCircularTextAnglePositionPanel() {
    return SliverToBoxAdapter(
      child: Row(
        mainAxisAlignment: MainAxisAlignment.spaceBetween,
        children: <Widget>[
          Text("START ANGLE ALIGNMENT",
              style: TextStyle(fontWeight: FontWeight.bold)),
          DropdownButton<StartAngleAlignment>(
            value: _startAngleAlignment,
            items: [
              DropdownMenuItem(
                child: Text("START"),
                value: StartAngleAlignment.start,
              ),
              DropdownMenuItem(
                child: Text("CENTER"),
                value: StartAngleAlignment.center,
              ),
              DropdownMenuItem(
                child: Text("END"),
                value: StartAngleAlignment.end,
              ),
            ],
            onChanged: (dynamic value) {
              setState(() => _startAngleAlignment = value);
            },
          )
        ],
      ),
    );
  }

  Widget buildStrokeWidthPanel() {
    return SliverToBoxAdapter(
      child: Row(
        mainAxisAlignment: MainAxisAlignment.spaceBetween,
        children: <Widget>[
          Text("STROKE WIDTH", style: TextStyle(fontWeight: FontWeight.bold)),
          Slider(
            value: _strokeWidth,
            min: 0,
            max: 100,
            onChanged: _showStroke
                ? (value) {
                    setState(() => _strokeWidth = value);
                  }
                : null,
          ),
          Checkbox(
            value: _showStroke,
            onChanged: (dynamic value) {
              setState(() => _showStroke = value);
            },
          )
        ],
      ),
    );
  }

  Widget buildShowBackgroundShapePanel() {
    return SliverToBoxAdapter(
      child: Row(
        mainAxisAlignment: MainAxisAlignment.spaceBetween,
        children: <Widget>[
          Text("SHOW BACKGROUND",
              style: TextStyle(fontWeight: FontWeight.bold)),
          Checkbox(
            value: _showBackground,
            onChanged: (dynamic value) {
              setState(() => _showBackground = value);
            },
          )
        ],
      ),
    );
  }

  Widget buildCircularTextPositionPanel() {
    return SliverToBoxAdapter(
      child: Row(
        mainAxisAlignment: MainAxisAlignment.spaceBetween,
        children: <Widget>[
          Text("POSITION", style: TextStyle(fontWeight: FontWeight.bold)),
          DropdownButton<CircularTextPosition>(
            value: _position,
            items: [
              DropdownMenuItem(
                child: Text("INSIDE"),
                value: CircularTextPosition.inside,
              ),
              DropdownMenuItem(
                child: Text("OUTSIDE"),
                value: CircularTextPosition.outside,
              )
            ],
            onChanged: (dynamic value) {
              setState(() => _position = value);
            },
          )
        ],
      ),
    );
  }

  Widget buildCircularTextDirectionPanel() {
    return SliverToBoxAdapter(
      child: Row(
        mainAxisAlignment: MainAxisAlignment.spaceBetween,
        children: <Widget>[
          Text("DIRECTION", style: TextStyle(fontWeight: FontWeight.bold)),
          DropdownButton<CircularTextDirection>(
            value: _direction,
            items: [
              DropdownMenuItem(
                child: Text("CLOCKWISE"),
                value: CircularTextDirection.clockwise,
              ),
              DropdownMenuItem(
                child: Text("ANTI CLOCKWISE"),
                value: CircularTextDirection.anticlockwise,
              )
            ],
            onChanged: (dynamic value) {
              setState(() => _direction = value);
            },
          )
        ],
      ),
    );
  }
}


example/lib/multi_text_example.dart
import 'package:flutter/material.dart';
import 'package:flutter_circular_text/circular_text.dart';

class MultiTextDemo extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Material(
      child: Container(
        color: Color(0xFF303030),
        alignment: Alignment.center,
        child: Stack(
          alignment: Alignment.center,
          children: <Widget>[
            _buildBackground(),
            _buildCircularTextWidget(),
          ],
        ),
      ),
    );
  }

  Widget _buildCircularTextWidget() {
    return CircularText(
      children: [
        TextItem(
          text: Text(
            "Chuck Norris".toUpperCase(),
            style: TextStyle(
              fontSize: 28,
              color: Colors.white,
              fontWeight: FontWeight.bold,
            ),
          ),
          space: 12,
          startAngle: -90,
          startAngleAlignment: StartAngleAlignment.center,
          direction: CircularTextDirection.clockwise,
        ),
        TextItem(
          text: Text(
            "top 100 Facts".toUpperCase(),
            style: TextStyle(
              fontSize: 20,
              color: Colors.amberAccent,
              fontWeight: FontWeight.bold,
            ),
          ),
          space: 10,
          startAngle: 90,
          startAngleAlignment: StartAngleAlignment.center,
          direction: CircularTextDirection.anticlockwise,
        ),
        TextItem(
          text: Text(
            "༒".toUpperCase(),
            style: TextStyle(
              fontSize: 18,
              color: Colors.white,
              fontWeight: FontWeight.bold,
            ),
          ),
          space: 10,
          startAngle: 180,
          startAngleAlignment: StartAngleAlignment.center,
          direction: CircularTextDirection.clockwise,
        ),
        TextItem(
          text: Text(
            "༒".toUpperCase(),
            style: TextStyle(
              fontSize: 18,
              color: Colors.white,
              fontWeight: FontWeight.bold,
            ),
          ),
          space: 10,
          startAngle: 0,
          startAngleAlignment: StartAngleAlignment.center,
          direction: CircularTextDirection.clockwise,
        ),
      ],
      radius: 135,
      position: CircularTextPosition.inside,
    );
  }

  Widget _buildBackground() {
    return Container(
      width: 300,
      height: 300,
      decoration: BoxDecoration(shape: BoxShape.circle, boxShadow: [
        BoxShadow(
          color: Color(0xFF101010),
        ),
        BoxShadow(
          color: Color(0xFF303030),
          blurRadius: 10,
        ),
      ]),
    );
  }
}
